Azure eventhub put processor enhancement

cima opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

Commit:
Batch & Partitioning key

+ Asynchronous sending in limited batches
+ Property to determine attribute name carrying partitioning key
+ Maximum batch size property
+ Carrier object
- Unit test fakery

Commit:
Attributes to User Defined properties

Commit:
Unit tests

arides commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/PutAzureEventHub.java:
```diff
@@ -150,6 +179,127 @@ public void tearDown() {
+			getLogger().error("Batch processing failed", e);
+			session.rollback();
+			throw new ProcessException("Batch processing failed", e);
+		}
```
For `InterruptedException`, add `Thread.currentThread().interrupt()` to avoid "swallowing" the interrupt for the rest of the application.

cima replied:
With respect to https://github.com/apache/nifi/pull/115/commits/a15b44485f4f062f0bfae5bed70c9566b5932649 I choose to proceed with your proposal.

arides commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/PutAzureEventHub.java:
```diff
@@ -206,22 +342,44 @@ protected EventHubClient createEventHubClient(
+     * 
+     * {@link https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-quotas} 
+     */
+    protected CompletableFuture<Void> sendMessage(final byte[] buffer, String partitioningKey, Map<String, ? extends Object> userProperties) throws ProcessException {
```
Is there are reason to explicitly use `<? extends Object>` instead of just `<?>`? 
If not, consider using the shorter version.

cima replied:
Everywhere Just `Object` is needed. The generic was automatically generated by refactoring tools. I choos to have `Map<String, Object>`

arides commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/FlowFileResultCarrier.java:
```diff
@@ -0,0 +1,41 @@
+
+	private FlowFile flowFile;
+	private T result;
+	private Throwable exception;
```
Consider making the fields final and removing setters, since it seems all of the fields fields are only ever meant to be filled when this object is created.

Commit:
Review corrections

+ Interruption propagation (& test)
+ Final carrier members
+ Unnecessary generic modifiers removed from generic container
